### ISSUE TITLE
Reduce repeated code for child classes of BaseStore

### ISSUE THREAD

**Issue by brendan-m-murphy** [openghg/openghg#727]
### What is your issue?

The classes that inherit from BaseStore have a lot of similar code that is meant to perform similar functions.

BaseStore helps the child classes interact with the object store, but the logic for standardising files is contained in a method called `read_file` in each child class.

There are a few issues here:
1. Code is duplicated between child classes.
2. Some features are only implement for some child classes (e.g. only `standardise_surface` checks that the metadata sent to the object store agrees with the metadata saved to the attributes of the netCDF data file).
3. The main purpose of BaseStore seems to be interacting with the object store, while the purpose of the child classes is processing data and metadata (then calling a method inherited from BaseStore to save the data and metadata to the object store). 

My plan is to make an abstract base class with functions for the various steps involved in standardised data. The common elements of the process will be in this class, and then for each data type, we can add to or override the base methods. Ideally it would work out so that in the child classes, you should just see that parts that are specific to that class. (ABC might not be the only option... I'll see what other options there are for a uniform interface.)

Initially, this ABC will be separate from BaseStore, and we can decide if we want to incorporate it into BaseStore, or split BaseStore out into separate classes for data processing and for communicating with the object store.

**Comment by brendan-m-murphy** [comment]
I've realised that the issue of duplicated processing code is not entirely an issue with children of BaseStore.

The `Footprints` child of `BaseStore` is somewhat unusual because there isn't code in the "standardise" module for standardising footprints; instead all of this code is in the `read_file` method of footprints.

Generally, I think the steps that need to happen are:
1. Take raw data and metadata (from input file and from user input), and return one metadata dictionary, and one data file in satisfying the OpenGHG schema for that data type. (Or a list of (metadata, data) pairs, if there are multiple files.)
2. Process the metadata from the previous step, make attributes CF compliant, make sure metadata and attributes are consistent. Write any new/changed attributes the data file. (This step returns a (metadata, data) pair.)
3. Add the data to the object store based on the metadata (including lookups for data sources, etc.).

It seems like it is desirable to check file hashes early on; maybe this can be managed by the children of `BaseStore`.

For surface data, attributes are checked for CF compliance in the parsing functions (at least for some parsing functions I've looked at). I think this check also happens when metadata and attributes are compared. Also attributes are added to the data in the parse function, but also in `assign_data` in the `BaseStore` class.

There is definitely a lot of variation in how step 1 is carried out, and that's inevitable, but if we can make it so that all the processing that is specific to certain data formats happens before other processing, and that the output of that step is uniform, then the rest of the code should be easier to streamline.

**Comment by rt17603** [comment]
Yeah you've spotted another inconsistency in our data type classes - only surface and emissions link to specific parse function at the moment. Again, emissions could be a blueprint for splitting this out - unlike surface the only available parse function is [parse_openghg](https://github.com/openghg/openghg/blob/devel/openghg/standardise/emissions/_openghg.py).

**Comment by rt17603** [comment]
This may be outside the scope of this issue for the moment but I also wanted to point to Issue #351 as this may be something to consider when re-writing this. The idea here was to allow users to pass alternative parse functions to the standardise --> .read_file() workflow which took the same inputs and produced the same style output as our in-built parse functions. The idea was to allow user customisation - as long as their outputs matched to our standardised format, they could add to the object store.

If we could have an eye on that with the design of this, that would be helpful.

**Comment by brendan-m-murphy** [comment]
The `parse_openghg` functions seems a bit out of place to me. They mostly process metadata, whereas other parse functions have to read various file formats. Maybe formatting metadata should be separate from "standardising" disparate sources to the OpenGHG standard. Then the parse function can just be responsible for passing a xarray Dataset and a dictionary of metadata, and a later step can be responsible for formatting the metadata and attributes. (And if the source format is openghg, then parsing isn't really needed.)

**Comment by brendan-m-murphy** [comment]
Also I think maybe `BaseStore` is better described as an "object store connection" or something, since it holds info about where to find the metastore and so on.

The `read_file` functions only exist in the child classes, and it might be better to keep them separate from functions/classes that enable reading/writing to the object store. I think this also means that the difference between local object stores and cloud object stores could possibly just depend on what type of "object store connection" you pass to the standardisation process. As long as the interface for "object store connection" allows for querying the metastore, retrieving files, checking hashes, etc, it doesn't really matter (for standardising) how it is implemented.

**Comment by rt17603** [comment]
> The `parse_openghg` functions seems a bit out of place to me. They mostly process metadata, whereas other parse functions have to read various file formats. Maybe formatting metadata should be separate from "standardising" various source formats to the OpenGHG standard. Then the parse function can just be responsible for passing a xarray Dataset and a dictionary of metadata, and a later step can be responsible for formatting the metadata and attributes. (And if the source format is openghg, then parsing isn't really needed.)

The idea with `parse_openghg` is that this can be very light touch but fits better into a more expandable structure e.g. as highlighted in issue https://github.com/openghg/openghg/issues/351 . By using a function such as this we set ourselves up to allow for other formats to be include.

**Comment by rt17603** [comment]
And I would also say that it would probably be best to attack this in logical stages rather than trying to rework the whole thing all at once. What steps could you see this be broken up into do you think?

**Comment by brendan-m-murphy** [comment]
Some steps could be:
- Change other data classes (footprints, boundary conditions, etc) to use standardise module, like emissions and surface
- Try to create a metadata class that collects and a formats metadata (hopefully to reduce code duplication, maybe with subclasses for different data types)
- Try creating a "object store connection" module to split communicating with the object store from formatting data. (I expect this would be essentially a subset of BaseObject, and could have child classes.) What is left over should bridge parsing and writing to the object store (depending on the data type, this might not include much at all, e.g. for emissions).

**Comment by brendan-m-murphy** [comment]
> The idea with `parse_openghg` is that this can be very light touch but fits better into a more expandable structure e.g. as highlighted in issue #351 . By using a function such as this we set ourselves up to allow for other formats/parse functions to be included.
> 
> I also think the metadata is so intertwined with the input details it makes sense to me to keep together the data and metadata extraction in one logical place even if some steps of this are passed off to other functions.

I think metadata/attribute formatting/comparing/assigning can't rely on the output of the parsers. Other wise, we need to make sure each parser is correct. So if we need to do the processing later, then the metadata formatting in the parser stage can be fairly minimal. (E.g. regardless of what data your parser reads in, we expect a single xr.Dataset and a single metadata dictionary out.)

I guess I'm thinking that if the format is "openghg", then you don't need to do step 1 (above), so you go straight to metadata/attribute formatting.

On the other hand, having parse_openghg as a model for similar parsers could be helpful.

**Comment by brendan-m-murphy** [comment]
Related issue: make sure `read_file` output format consistent: https://github.com/openghg/openghg/issues/268

**Comment by rt17603** [comment]
May be related to: #669

**Comment by brendan-m-murphy** [comment]
As part of this update either:
1) move `object_store_data_classes` to a central location like `store.spec`, or
2) ensure that the only references to `object_store_data_classes` are in the `objectstore` module

https://github.com/openghg/openghg/pull/771#discussion_r1320122014

**Comment by brendan-m-murphy** [comment]
Inconsistencies around `_metakey` that should be addressed (perhaps as part of the comments from 13 Sep. 2023):

_metakey is a class variable in the subclasses of BaseStore but an instance variable for BaseStore itself. I think maybe _metakey shouldn't be a class variable, and the class method metakey should make the metakey using the class variables _root and _uuid. Maybe this was something that was going to be changed when I was refactoring the children of BaseStore

**Comment by rt17603** [comment]
> Inconsistencies around `_metakey` that should be addressed (perhaps as part of the comments from 13 Sep. 2023):
> 
> _metakey is a class variable in the subclasses of BaseStore but an instance variable for BaseStore itself. I think maybe _metakey shouldn't be a class variable, and the class method metakey should make the metakey using the class variables _root and _uuid. Maybe this was something that was going to be changed when I was refactoring the children of BaseStore

New issue opened for this #1537
